File: ScanProject/JavaAnalysis/SonarJava/src/main/java/Medium/663-subsets-ii.java
package main.java.Medium;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**https://leetcode.com/problems/subsets-ii/ */
//Given an integer array nums that may contain duplicates, return all possible subsets (the power set).
//The solution set must not contain duplicate subsets. Return the solution in any order.
// 
//Example 1:
//Input: nums = [1,2,2]
//Output: [[],[1],[1,2],[1,2,2],[2],[2,2]]
//Example 2:
//Input: nums = [0]
//Output: [[],[0]]
// 
//Constraints:
//	1 <= nums.length <= 10
//	-10 <= nums[i] <= 10
class Solution454 {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        boolean[] visited = new boolean[nums.length];
        Arrays.sort(nums);
        dfs(nums, 0, visited, new ArrayList<>(), res);
        return res;
    }

    public void dfs(int[] nums, int start, boolean[] visited, List<Integer> path, List<List<Integer>> res) {
        res.add(new ArrayList<>(path));
        for (int i = start; i < nums.length; i++) {
            // key point: if the former one is not visited, and the former one is the same as the current one, then skip the current one
            if (i > start && nums[i - 1] == nums[i] && !visited[i - 1]) {
                continue;
            }
            path.add(nums[i]);
            visited[i] = true;
            dfs(nums, i + 1, visited, path, res);
            visited[i] = false;
            path.remove(path.size() - 1);
        }
    }
}

File: ScanProject/PyAnalysis/SonarPy/Medium/663-subsets-ii.py
#https://leetcode.com/problems/subsets-ii/
#Given an integer array nums that may contain duplicates, return all possible subsets (the power set).
#The solution set must not contain duplicate subsets. Return the solution in any order.


class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def check(res, item):
            for i in res:
                if len(i) != len(item):
                    continue
                for j in range(len(i)):
                    if i[j] != item[j]:
                        break
                else:
                    return False
            return True
        def dfs(nums, start, item, res):
            if check(res, item):
                res.append(item[:])
            for i in range(start, len(nums)):
                item.append(nums[i])
                dfs(nums, i + 1, item, res)
                item.pop()
        nums.sort()
        res = []
        dfs(nums, 0, [], res)
        return res
        

    